### DISCUSSION TITLE
Using vd for json/yaml (bibliography)

### DISCUSSION THREAD

**Discussion by reagle** [saulpw/visidata#1476] [Q&A] [answered]
I'm thrilled that I can use vd to view json/yaml, which are the formats I keep my bibliographies in. However, I'm don't understand how  vd presents nested structures. For example, in the following I think it is telling me there are 5 authors, but what are the `{2}`s? That there is a nesting for first and last name?

<img width="641" alt="Screen Shot 2022-08-12 at 10 20 31" src="https://user-images.githubusercontent.com/596814/184373258-81e63aaf-89b5-4aa2-94b6-acafe44c61c1.png">

I'm also wondering how to query/filter, on dates, for example:

<img width="652" alt="Screen Shot 2022-08-12 at 10 22 17" src="https://user-images.githubusercontent.com/596814/184373618-65b37dbc-820a-4a9e-b9bf-98e53eca2967.png">

Setting it to a date type `@` doesn't work as I expected. Is anyone using visidata in this way and have suggestions (e.g., find all entries with author X and published between Y and Z)?

**Comment by saulpw** [comment, marked as answer]
Hey @reagle, yes, `[2]` and `{2}` represent container objects (list and dict, respectively).  You probably want to use `(` to expand dicts into multiple columns.  For lists you can use `(` also if you want them columnarly, or `zM` to unfurl them into multiple rows.

For dates, python-dateutil understands a lot of formats, but it doesn't understand this `date-parts` list object yet :)  You'll have to make a new column (probably using `=`) that formats it into a date string, and then you can use `@` on that.

**Reply by saulpw** [reply]
BTW is this a common bibliography format, or a custom one that you came up with?

**Reply by reagle** [reply]
Okay, that information is helpful. After trying unfurl and melt (which I read about in the tutorial), I don't yet understand them, so I'll have to tinker some more.

The bibliography formats are CSL JSON (and CSL YAML, which is much slower to parse understandably) specified by the [Citation Style Language (CSL)](https://citeproc-js.readthedocs.io/en/latest/csl-json/markup.html) and used by [pandoc](https://pandoc.org/) and other citeproc apps.

**Comment by geekscrapy** [comment]
> <img alt="Screen Shot 2022-08-12 at 10 20 31" width="641" src="https://user-images.githubusercontent.com/596814/184373258-81e63aaf-89b5-4aa2-94b6-acafe44c61c1.png">

So, in this example the underlying data is like?:

`[{family:v1, given: v2},{family:v1, given: v2}]`

FWIW @spw this is the kind of formatting I was referring to a while back. I find a little hard to understand at first glance and I think you really need to know a bit of python to fully understand it

**Reply by reagle** [reply]
Here's the example in CSL JSON:

    { "id": "ReagleEtal1999apu",
      "type": "report",
      "author": [
          { "family": "Reagle", "given": "Joseph M." },
          { "family": "Weitzner", "given": "Daniel J." },
          { "family": "Rein", "given": "Barry D." },
          { "family": "Stephens", "given": "Garland T." },
          { "family": "Lebowitz", "given": "Henry C." }
        ],
      "custom2": "pubs.mm",
      "issued": {"date-parts": [ [ 1999, 10 ] ]
      },
      "publisher": "W3C",
      "title": "Analysis of P3P and US Patent 5,862,325",
      "genre": "Note",
      "URL": "http://www.w3.org/TR/P3P-analysis"
}

**Reply by reagle** [reply]
BTW: How can I easily search an `author` field that is a 3 items list (`[3]{2})? I think the `/` only matches the last name of the first item?

**Reply by saulpw** [reply]
The easiest way is set the column type to `str` (with `~`), and then search within that.  You can use `z~` to change the type back to `anytype`.
